pkg/jobWorker/job.go

```diff
@@ -0,0 +1,307 @@
+	if err = ns.AddProcess(cgroupName, cmd); err != nil {
+		log.Printf("Error AddProcess /proc - %s\n", err)
+		os.Exit(1)
+	}
+
+	if err = ns.MountProc(); err != nil {
+		log.Printf("Error mounting /proc - %s\n", err)
+		os.Exit(1)
+	}
+
+	log.Printf("starting job:%s, cmd:%s", job, cmd.String())
+	if err = cmd.Start(); err != nil {
+		return fmt.Errorf("error starting command: %w", err)
+	}
+	job.cmd = cmd
+	job.status.State = JOB_STATUS_RUNNING
+
+	// run the command in a Goroutine so that Start can return immediately
+	go func() {
+		// Use cmd.Process.Wait() instead of cmd.Wait() since cmd.Wait() is not thread safe
+		// and we do not want to hold the mutex while waiting for the process to exit.
+		// So instead we use cmd.Process.Wait() and store the result in j.processState to mimic what
+		// cmd.Wait would do.
+		// This prevents concurrency issues when a user calls Start(), the command quickly exits (updating the
+		// process state), and the user invokes Status().
+		processState, err := job.cmd.Process.Wait()
+
+		job.mutex.Lock()
+		defer job.mutex.Unlock()
+
+		job.processState = processState
+		job.status.ExitCode = job.processState.ExitCode()
+		if job.status.State == JOB_STATUS_RUNNING {
+			job.status.State = JOB_STATUS_COMPLETED
+		}
+
+		if err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", err)
+		}
+
+		if err == nil && !job.processState.Success() {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", &exec.ExitError{ProcessState: job.processState})
+		}
+
+		// close the output, so that any readers of the output know the process has exited and will no longer
+		// block waiting for new output
+		if err = job.output.Close(); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing output: %s\n", err)
+		}
+
+		// do not close the cgroup.procs file until after the process has exited
+		if err = ns.DeleteCGroup(cgroupName); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing cgroup: %s\n", err)
+		}
+
+		if err = ns.UnmountProc(); err != nil {
+			log.Printf("error unmounting /proc - %s\n", err)
+		}
+	}()
+
+	return nil
+}
+
+// Status returns the current Status of the Job.
+func (job *Job) Status() *JobStatus {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+	log.Printf("get job status:%s", job)
+	return job.status
+}
+
+// Stream returns an OutputReadCloser that streams the combined stdout and stderr of the Job.
+func (job *Job) Stream() *OutputReadCloser {
+	log.Printf("get job stream:%s", job)
+	return NewOutputReadCloser(job.output)
+}
+
+func (job *Job) Stop() error {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+
+	log.Printf("stop job :%s", job)
+	if job.cmd == nil {
+		return ErrJobNotStarted
+	}
+
+	if err := job.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+		return fmt.Errorf("error sending SIGTERM: %w", err)
+	}
```

Comment:
What if the process has children. Does the job.cmd.Process.Signal will also stop/terminate all process children's  ?

Reply:
@smallinsky  That really interesting question. After research I've found we just need to send negative PID to KILL all child process. See here: https://linux.die.net/man/2/kill
`
If pid equals -1, then sig is sent to every process for which the calling process has permission to send signals, except for process 1 (init), but see below.

If pid is less than -1, then sig is sent to every process in the process group whose ID is -pid. 
`
And changes would looks like 
<img width="1102" alt="Screenshot 2024-09-04 at 2 23 17 PM" src="https://github.com/user-attachments/assets/0d106b93-3524-450d-be0c-4e06ec0d4c37">

But such things should be deeply tested. What do you think ?

Reply:
@smallinsky yes, changed. It is much elegance solutions than mine.  
Thank you.